diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/StringConcatenationRule.kt

```diff
@@ -36,25 +49,56 @@ class StringConcatenationRule(configRules: List<RulesConfig>) : DiktatRule(
 
-    private fun detectStringConcatenation(node: ASTNode) {
-        node.findAllNodesWithSpecificType(BINARY_EXPRESSION).find { detectStringConcatenationInExpression(it, node) }
+    /**
+     * This method works only with top-level binary expressions. It should be checked before the call.
+     */
+    @Suppress("AVOID_NULL_CHECKS")
+    private fun detectStringConcatenation(topLevelBinaryExpr: ASTNode) {
+        val allBinaryExpressions = topLevelBinaryExpr.findAllDescendantsWithSpecificType(BINARY_EXPRESSION)
+        val nodeWithBug = allBinaryExpressions.find { isDetectStringConcatenationInExpression(it) }
+
+        if (nodeWithBug != null) {
+            STRING_CONCATENATION.warnAndFix(
+                configRules, emitWarn,
+                this.isFixMode, topLevelBinaryExpr.text.lines().first(), nodeWithBug.startOffset, nodeWithBug
+            ) {
+                fixBinaryExpressionWithConcatenation(nodeWithBug)
+                loop(topLevelBinaryExpr.treeParent)
+            }
+        }
     }
 
-    @Suppress("FUNCTION_BOOLEAN_PREFIX")
-    private fun detectStringConcatenationInExpression(node: ASTNode, parentNode: ASTNode): Boolean {
-        assert(node.elementType == BINARY_EXPRESSION)
-        val firstChild = node.firstChildNode
-        return if (isPlusBinaryExpression(node) && firstChild.elementType == STRING_TEMPLATE) {
-            STRING_CONCATENATION.warn(configRules, emitWarn, this.isFixMode, parentNode.text, firstChild.startOffset, firstChild)
-            true
-        } else {
-            false
+    private fun loop(parentTopLevelBinaryExpr: ASTNode) {
+        val allBinaryExpressions = parentTopLevelBinaryExpr.findAllDescendantsWithSpecificType(BINARY_EXPRESSION)
+        val nodeWithBug = allBinaryExpressions.find { isDetectStringConcatenationInExpression(it) }
+
+        val bugDetected = nodeWithBug != null
+        if (bugDetected) {
+            fixBinaryExpressionWithConcatenation(nodeWithBug)
+            loop(parentTopLevelBinaryExpr)
+        }
+    }
+
+    /**
+     * We can detect string concatenation by the first (left) operand in binary expression.
+     * If it is of type string - then we found string concatenation.
+     * If the right value is not a constant string then don't change them to template.
+     */
+    private fun isDetectStringConcatenationInExpression(node: ASTNode): Boolean {
+        require(node.elementType == BINARY_EXPRESSION) {
+            "cannot process non binary expression in the process of detecting string concatenation"
         }
+        val firstChild = node.firstChildNode
+        val lastChild = node.lastChildNode
+        return (isPlusBinaryExpression(node) && !(firstChild.elementType != STRING_TEMPLATE && lastChild.elementType == CALL_EXPRESSION) &&
+                (firstChild.elementType == STRING_TEMPLATE ||
+                        ((firstChild.text.endsWith("toString()")) && firstChild.elementType == DOT_QUALIFIED_EXPRESSION))
```

Comment:
```suggestion
        return (isPlusBinaryExpression(node) && (firstChild.elementType == STRING_TEMPLATE ||
                        (firstChild.text.endsWith("toString()") && firstChild.elementType == DOT_QUALIFIED_EXPRESSION && lastChild.elementType != CALL_EXPRESSION))
```
I'd merge these two conditions for readability. Also, it might be worth extracting them to a variable.